src/js/components/OptionalEnviromentComponent.jsx

```diff
@@ -0,0 +1,99 @@
+var React = require("react/addons");
+
+var appValidator = require("../validators/appValidator");
+var FormGroupComponent = require("../components/FormGroupComponent");
+var DuplicableRowControls = require("../components/DuplicableRowControls");
+
+var OptionalSettingsComponent = React.createClass({
+  displayName: "OptionalEnvironmentComponent",
+
+  propTypes: {
+    errors: React.PropTypes.array,
+    model: React.PropTypes.object.isRequired
+  },
+
+  getInitialState: function () {
+    var env = this.props.model.env;
+    var state = {
+      env: Object.keys(env).map(function (key) {
+        return {
+          key: key,
+          value: env[key]
+        };
+      })
+    };
+    if (state.env.length === 0) {
+      state.env = [true];
+    }
+    return state;
+  },
+
+  handleAddRow: function (position, event) {
+    event.target.blur();
+    event.preventDefault();
+    var env = this.state.env.concat(true);
```

Comment:
+1 on the use of `concat`, which does not mutate the state directly. 
To keep things more React-y you could have used the update syntax from the [immutability helpers](https://facebook.github.io/react/docs/update.html), but `concat` does keep things short ;)